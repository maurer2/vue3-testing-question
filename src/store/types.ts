import {
  ActionContext, MutationPayload, Store as GenericStore, ActionPayload, ActionTree,
} from 'vuex';

const appStates = ['initial', 'loaded', 'error'] as const;
const actionsMap = {
  updateNumberOfClicks: 'UPDATE_NUMBER_OF_CLICKS',
  test: 'TEST',
} as const;

type ActionsName = keyof typeof actionsMap

type State = {
  appState: typeof appStates[number]
  numberOfClicks: number,
}

type Getters = {
  HAS_BEEN_CLICKED(state: State): boolean
}

type Actions<S = State> = {
  // actionNames['UPDATE_NUMBER_OF_CLICKS'],

  UPDATE_NUMBER_OF_CLICKS(context: ActionContext<S, S>, payload: ActionPayload): Promise<void>;
};

type Mutations<S = State> = {
  UPDATE_NUMBER_OF_CLICKS(state: S, payload: MutationPayload): void;
};

type Store = GenericStore<State>

export {
  appStates,
  actionsMap,
  Store,
  Actions,
  State,
  Getters,
  Mutations,
};
