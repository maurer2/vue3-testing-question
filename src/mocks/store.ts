import { State } from '../store/types';

const storeInitial = {
  state: {
    appState: 'initial',
    numberOfClicks: 0,
  } as State,
  dispatch: jest.fn(),
};

export {
  storeInitial,
};
