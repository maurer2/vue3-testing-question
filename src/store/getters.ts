import { GetterTree } from 'vuex';
import { State } from './types';

const getterNames = Object.freeze({
  hasBeenClicked: 'HAS_BEEN_CLICKED',
} as const);

type Getters = {
  [getterNames.hasBeenClicked](state: State): boolean
}

const getters: GetterTree<State, State> & Getters = {
  [getterNames.hasBeenClicked](state) {
    return state.numberOfClicks !== 0;
  },
};

export {
  getters,
};

export type {
  Getters,
};
