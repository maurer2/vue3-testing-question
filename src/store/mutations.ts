import { MutationTree, MutationPayload } from 'vuex';
import { State } from './types';

const mutationNames = Object.freeze({
  UPDATE_NUMBER_OF_CLICKS: 'UPDATE_NUMBER_OF_CLICKS',
} as const);

const mutations: MutationTree<State> = {
  [mutationNames.UPDATE_NUMBER_OF_CLICKS](state, payload?: Readonly<MutationPayload>): void {
    // eslint-disable-next-line no-param-reassign
    state.numberOfClicks += 1;
  },
};

export {
  mutations,
  mutationNames,
};
