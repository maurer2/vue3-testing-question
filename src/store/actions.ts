import { ActionTree, ActionPayload } from 'vuex';
import { State } from './types';
import { mutationsNames } from './mutations';

const actionsNames = Object.freeze({
  updateNumberOfClicks: 'UPDATE_NUMBER_OF_CLICKS',
} as const);

type ActionsName = keyof typeof actionsNames

const actions: ActionTree<State, State> = {
  [actionsNames.updateNumberOfClicks](context, payload?: ActionPayload): Promise<void> {
    context.commit(mutationsNames.updateNumberOfClicks);

    return Promise.resolve();
  },
};

export {
  actions,
  actionsNames,
};

export type {
  ActionsName,
};
