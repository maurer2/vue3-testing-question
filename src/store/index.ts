/* eslint-disable no-param-reassign */
import {
  createStore, useStore as genericUseStore, ActionTree, ActionPayload,
} from 'vuex';
import { InjectionKey } from 'vue';

import {
  actionsName,
  appStates,
  // Actions,
  State,
  Store,
} from './types';

import { actions } from './actions';
import { mutations } from './mutations';
import { getters } from './getters';

const key: InjectionKey<Store> = Symbol('key');
const store = createStore<State>({
  state() {
    return {
      appState: 'loaded',
      numberOfClicks: 0,
    };
  },
  getters,
  actions,
  mutations,
});

export function useStore(): Store {
  return genericUseStore(key);
}

export {
  store,
  key,
};
