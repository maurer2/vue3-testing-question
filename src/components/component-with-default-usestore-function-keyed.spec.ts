// eslint-disable-next-line import/no-extraneous-dependencies
import { shallowMount } from '@vue/test-utils';
import { nanoid } from 'nanoid';
import Component from './component-with-default-usestore-function-keyed.vue';

import { storeInitial as storeMock, key } from '../mocks/store';
import { store as storeReal } from '../store/index';

jest.mock('nanoid', () => ({
  nanoid: jest.fn().mockReturnValue('button-keyed'),
}));

jest.mock('../mocks/store.ts', () => ({
  ...jest.requireActual('../mocks/store.ts'),
  // dispatch: jest.fn(),
}));

describe('App', () => {
  let cmp: any;

  beforeEach(() => {
    cmp = shallowMount(Component, {
      global: {
        provide: {
          [key as symbol]: storeMock,
          // [key as symbol]: storeReal,
        },
      },
    });

    jest.spyOn(cmp.vm, 'handleClick');
    jest.spyOn(storeMock, 'dispatch');
  });

  afterEach(() => {
    cmp.unmount();

    jest.clearAllMocks();
    jest.resetModules();
  });

  it('renders', () => {
    expect(cmp.html()).toContain('Click');
  });

  it('matches snapshot', () => {
    // https://github.com/eddyerburgh/jest-serializer-vue/issues/49
    // eslint-disable-next-line no-underscore-dangle
    expect(cmp.__app._container).toMatchSnapshot();
  });

  it('button is correctly wired up with handleClick', async () => {
    expect(cmp.find('#id-button-keyed').exists()).toBe(true);

    await cmp.find('#id-button-keyed').trigger('click');
    expect(cmp.vm.handleClick).toHaveBeenCalled();
  });

  it('handleClick dispatches click update action', () => {
    // const spiedDispatch = jest.spyOn(storeMock, 'dispatch');
    cmp.vm.handleClick();

    // expect(spiedDispatch).toHaveBeenCalled();
    expect(storeMock.dispatch).toHaveBeenCalledWith('UPDATE_NUMBER_OF_CLICKS');
  });

  it('dispatches action updates the store', async () => {
    expect(storeMock.state.numberOfClicks).toBe(0);

    cmp.vm.handleClick();
    await expect(storeMock.dispatch).toHaveBeenCalledWith('UPDATE_NUMBER_OF_CLICKS');

    expect(storeMock.state).toBe(true);
  });
});
