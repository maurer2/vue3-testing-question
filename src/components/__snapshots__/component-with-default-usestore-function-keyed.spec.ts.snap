// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<div
  data-v-app=""
>
  <section>
    <button
      id="id-button-keyed"
      type="button"
    >
       Click 
    </button>
    <div>
       Debug 
      <code>
        {
  "appState": "loaded",
  "numberOfClicks": 0
}
      </code>
       - 
      <code>
        false
      </code>
       - 
      <code>
        false
      </code>
    </div>
  </section>
</div>
`;
